### PR TITLE
add builtin function hex

### PR DESCRIPTION
**Pull request by XuHuaiyu**
#1430

### CODE REVIEW

**Comment by shenli** [comment]
@tiancaiamao PTAL

File: parser/parser.y [hunk 1]
```diff
@@ -174,6 +174,7 @@ import (
 	groupConcat	"GROUP_CONCAT"
 	hash		"HASH"
 	having		"HAVING"
+    hex         "HEX"
```
Comment (zimulala) [inline parser/parser.y]: Keep alignment.

File: evaluator/builtin_string.go [hunk 2]
```diff
@@ -458,6 +460,27 @@ func builtinLocate(args []types.Datum, _ context.Context) (d types.Datum, err er
 
 const spaceChars = "\n\t\r "
 
+// See http://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_hex
+func builtinHex(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+	switch args[0].Kind() {
+	case types.KindNull:
+		return d, nil
+	case types.KindString:
+		x, err := args[0].ToString()
```
Comment (tiancaiamao) [inline evaluator/builtin_string.go]: take a look at tidb/util/hack.String(),  sometimes type conversion can be avoid

Comment (XuHuaiyu) [inline evaluator/builtin_string.go]: OK

File: parser/parser.y (hunk 1 above)
Comment (XuHuaiyu) [inline parser/parser.y]: OK

File: parser/parser.y [hunk 3]
```diff
@@ -174,6 +174,7 @@ import (
 	groupConcat	"GROUP_CONCAT"
 	hash		"HASH"
 	having		"HAVING"
+  hex         "HEX"
```
Comment (tiancaiamao) [inline parser/parser.y]: indent...

Comment (XuHuaiyu) [inline parser/parser.y]: OK...
I'll take a look..

File: parser/scanner.l [hunk 4]
```diff
@@ -388,6 +388,7 @@ group		{g}{r}{o}{u}{p}
 group_concat	{g}{r}{o}{u}{p}_{c}{o}{n}{c}{a}{t}
 hash		{h}{a}{s}{h}
 having		{h}{a}{v}{i}{n}{g}
+hex			{h}{e}{x}
```
Comment (tiancaiamao) [inline parser/scanner.l]: ident

**Comment by tiancaiamao** [comment]
Ident is important, otherwise when file edited by others, it's confused to figure out what's changed.
rest LGTM

**Comment by XuHuaiyu** [comment]
I have changed all the problems.
@tiancaiamao

File: evaluator/builtin_string.go [hunk 5]
```diff
@@ -458,6 +461,28 @@ func builtinLocate(args []types.Datum, _ context.Context) (d types.Datum, err er
...
+		if err != nil {
+			return d, nil
```
Comment (shenli) [inline evaluator/builtin_string.go]: return d, errors.Trace(err)

File: evaluator/builtin_test.go [hunk 6]
```diff
@@ -91,6 +91,31 @@ func (s *testEvaluatorSuite) TestGreatestFunc(c *C) {
 	c.Assert(v.IsNull(), IsTrue)
 }
 
+func (s *testEvaluatorSuite) TestHexFunc(c *C) {
+	defer testleak.AfterTest(c)()
+
+	v, err := builtinHex(types.MakeDatums(12), nil)
```
Comment (shenli) [inline evaluator/builtin_test.go]: Add a test case with nil as input.

Comment (shenli) [inline evaluator/builtin_test.go]: And also a test case which will return error.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
I add wrong test cases

File: evaluator/builtin_string_test.go [hunk 7]
```diff
@@ -658,3 +658,40 @@ func (s *testEvaluatorSuite) TestTrim(c *C) {
...
+		{-12.5, "FFFFFFFFFFFFFFF3"},
+		{"12", "3132"},
+		{0x12, "12"},
+		{"", ""},
+	}
+
+	dtbl := tblToDtbl(tbl)
+
+	for _, t := range dtbl {
+		d, err := builtinHex(t["Input"], nil)
```
Comment (shenli) [inline evaluator/builtin_string_test.go]: t.Input, t.Expect

File: evaluator/builtin_string_test.go [hunk 8]
```diff
@@ -658,3 +658,40 @@ func (s *testEvaluatorSuite) TestTrim(c *C) {
...
+
+	wrong := []struct {
+		Input interface{}
+	}{
+		{nil},
+		{},
+	}
+	dwrong := tblToDtbl(wrong)
+	for _, t := range dwrong {
+		_, err := builtinHex(t["Input"], nil)
```
Comment (shenli) [inline evaluator/builtin_string_test.go]: t.Input

**Comment by shenli** [comment]
LGTM

**Pull request merged**